runtime/decimal.c

```diff
@@ -129,6 +131,32 @@ TaggedPtr _bal_decimal_neg(TaggedPtr tp) {
     return createDecimal(&d);
 }
 
+TaggedPtrPanicCode _bal_decimal_from_float(double val) {
+    TaggedPtrPanicCode result;
+    if (isnan(val)) {
+        result.panicCode = PANIC_INVALID_DECIMAL;
+        return result;
+    }
+    if (isinf(val)) {
+        result.panicCode = PANIC_ARITHMETIC_OVERFLOW;
+        return result;
+    }
+    result.panicCode = 0;
+    decQuad d;
+    if (val == 0.0) {
+        decQuadZero(&d);
+        result.ptr = createDecimal(&d);
+        return result;
+    }
+    char str[EMYG_DTOA_BUFFER_LEN];
+    emyg_dtoa_non_special(val, str);
+    decContext cx;
+    initContext(&cx);
+    decQuadFromString(&d, str, &cx);
```

Comment:
We need to think about what the right precision is and whether this will give it to us.

Reply:
https://github.com/ballerina-platform/nballerina/pull/724#issuecomment-981325798